test/functional/ui/incsubstitution_spec.lua

```diff
@@ -0,0 +1,379 @@
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      ^these ARE nothing                       |
+      with colorful text (ARE)                |
+      with colorful text (ARE)                |
+      1001 substitutions on 1001 lines        |
+    ]])
+
+    feed('i')
+    feed('example of insertion')
+
+    screen:expect([[
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      these ARE nothing                       |
+      with colorful text (ARE)                |
+      example of insertion^these ARE nothing   |
+      with colorful text (ARE)                |
+      with colorful text (ARE)                |
+      {6:-- INSERT --}                            |
+   ]])
+
+   feed('<esc>')
+   execute('set incsubstitute=2')
+   execute('1,1000s/ARE/nut/g')
+   feed(':%s/ARE/to')
+
+   screen:expect([[
+     with colorful text (to)                 |
+     these to nothing                        |
+     with colorful text (to)                 |
+     these to nothing                        |
+     with colorful text (to)                 |
+     these to nothing                        |
+     with colorful text (to)                 |
+     example of insertionthese to nothing    |
+     with colorful text (to)                 |
+     with colorful text (to)                 |
+     {11:[No Name] [+]                           }|
+      [1001]with colorful text ({12:to})          |
+      [1002]these {12:to} nothing                 |
+      [1003]with colorful text ({12:to})          |
+      [1004]these {12:to} nothing                 |
+      [1005]with colorful text ({12:to})          |
+      [1006]these {12:to} nothing                 |
+      [1007]with colorful text ({12:to})          |
+     {10:[inc_sub]                               }|
+     :%s/ARE/to^                              |
+   ]])
+  end)
+
+end)
```

Comment:
Missing tests:

1. `s@test@…`. I know that currently it only works with `/`, but this only means that tests should check that Neovim does not crash (or whatever) with `@` as delimiters, not that all tests should use `/`.
2. One of the livesub tests, but with all characters coming from `cnoremap / /` (i.e. remapped to themselves).
3. Livesub test with characters coming from mapping that moves the cursor (e.g. `:%s/ARE,S` with `cnoremap ,S \(\)<left><left>`).
4. Livesub tests with unfinished failing pattern which *fails* to compile like `foo\(`.
5. Livesub tests with unfinished failing replacement: e.g. `%s/ARE/\='to`.
6. Livesub tests with finished failing replacement: e.g. `%s/ARE/\=xxx_nonexistent_var_xxx`.
7. Livesub splitview tests with failing mapping: e.g. `%s/ARE/tox` with `cnoremap <expr> x capture('bwipeout!')[-1].'x'` (this one should produce an error because `bwipeout!` fails due to textlock).
8. Livesub splitview tests with mapping which temporary moves the cursor: e.g. `%s/ARE/tox` with `cnoremap <expr> x cursor(1, 1)[-1].'x'`.
9. Livesub tests with mapping which disables livesub: e.g. `%s/ARE/tox` with `cnoremap <expr> x capture('set incsubstitute=')[-1]`. Especially interesting what splitview livesub will do here.
10. Livesub tests with mapping which saves command line, opens a temporary split with new buffer, wipes it out and restores command line: e.g. `%s/ARE/tox` with `cnoremap x <C-\>eextend(g:, {'foo_savedcmdline': getcmdline()}).foo_savedcmdline<CR><C-c>:new<CR>:bw!<CR>:<C-r>=remove(g:, 'foo_savedcmdline')<CR>x`. Again, splitview livesub is especially interesting.
11. Tests which check what autocommands (out of Buf\*, Win\*, Syntax and FileType) are run when ics=splitview opens and closes a window.
12. If any are run when window is opened, then tests which check what if these autocommands close the window.
13. If any are run when window is opened, then tests which check what if these autocommands wipe out buffer without closing the window.
14. If any are run when window is opened, then tests which check what if these autocommands switch buffer without closing the window or wiping buffer out.